kite-tutorials/crunch/src/org/kitesdk/examples/tutorials/crunch/AggregateEvents.java

```diff
@@ -0,0 +1,152 @@
+
+  private static final Logger LOG = LoggerFactory.getLogger(AggregateEvents.class);
+
+  @Override
+  public int run(String[] args) throws Exception {
+	  
+    // Turn debug on while in development.
+    getPipeline().enableDebug();
+    getPipeline().getConfiguration().set("crunch.log.job.progress", "true");
+    
+    // Step 1. Load the dataset into the Pipeline.
+    Dataset<StandardEvent> eventsDataset = Datasets.load(
+        "dataset:hive:events", StandardEvent.class);
+
+	// If the dataset is empty, stop.
+    if (eventsDataset.isEmpty()) {
+      LOG.info("No records to process.");
+      return 0;
+    }
+
+    // Create a parallel collection from the Kite CrunchDatasets
+    // Source "events" dataset.
+    PCollection<StandardEvent> events = read(
+        CrunchDatasets.asSource(eventsDataset));
+
+    /* Step 2. Process the data.
+     * a. Create a session key by combining the user ID and session ID.
+     * b. Group together all events with the same session key.
+     * c. For each group, create a Session record as an Avro Specific object. 
+     */
+    PCollection<Session> sessions = events
+        .by(new GetSessionKey(), Avros.strings())
+        .groupByKey()
+        .parallelDo(new MakeSession(), Avros.specifics(Session.class));
+
+    // Step 3. Append the derived sessions to the Kite CrunchDatasets
+    // Target "sessions" dataset.
+    getPipeline().write(sessions,
+        CrunchDatasets.asTarget("dataset:hive:sessions"),
+        Target.WriteMode.APPEND);
+
+    return run().succeeded() ? 0 : 1;
+  }
+
+  private static class GetSessionKey extends MapFn<StandardEvent, String> {
+    @Override
+    public String map(StandardEvent event) {
+      // Create a key by combining the session id and user id
+      return event.getSessionId() + event.getUserId();
+    }
+  }
+
+  private static class MakeSession
+      extends DoFn<Pair<String, Iterable<StandardEvent>>, Session> {
+
+    // The process method iterates through a group of events that have
+    // the same sessionKey.
+    @Override
+    public void process(
+        Pair<String, Iterable<StandardEvent>> keyAndEvents,
+        Emitter<Session> emitter) {
+      final Iterator<StandardEvent> events = keyAndEvents.second().iterator();
+      if (!events.hasNext()) {
+        return;
+      }
+
+      // Initialize the values needed to create a session object for
+      // this group.
+      final StandardEvent firstEvent = events.next();
+      long startTime = firstEvent.getTimestamp();
+      long endTime = firstEvent.getTimestamp();
+      int numEvents = 1;
+
+      // Inspect each event in this session group. Track the earliest 
+      // timestamp (start time) and latest timestamp (end time). Keep a
+      // count of the events in this session group.
+      while (events.hasNext()) {
+        final StandardEvent event = events.next();
+        
+        // Reset the start time if the timestamp is earlier than the
```

Comment:
The word reset isn't quite appropriate here. It would be better to say "if the current timestamp is before the earliest seen so far, update the earliest."